staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go

```diff
@@ -143,22 +145,25 @@ func ErrorNegotiated(err error, s runtime.NegotiatedSerializer, gv schema.GroupV
 	return code
 }
 
-// errorJSONFatal renders an error to the response, and if codec fails will render plaintext.
+// errorSerializationFatal renders an error to the response, and if codec fails will render plaintext.
 // Returns the HTTP status code of the error.
-func errorJSONFatal(err error, codec runtime.Encoder, w http.ResponseWriter) int {
+func errorSerializationFatal(err error, originalStatusCode int, codec runtime.Encoder, w http.ResponseWriter) {
 	utilruntime.HandleError(fmt.Errorf("apiserver was unable to write a JSON response: %v", err))
 	status := ErrorToAPIStatus(err)
 	code := int(status.Code)
+	// If original statusCode was not successful, we need to return the original error.
+	// We cannot hide it behind serialization problems
+	if originalStatusCode < http.StatusOK || originalStatusCode >= http.StatusBadRequest {
+		code = originalStatusCode
+	}
+	w.WriteHeader(code)
 	output, err := runtime.Encode(codec, status)
 	if err != nil {
-		w.WriteHeader(code)
+		w.Header().Set("Content-Type", "text/plain")
```

Comment:
Should we modify header before calling WriteHeader?

From the document: https://golang.org/pkg/net/http/#ResponseWriter
```
        // Changing the header map after a call to WriteHeader (or
        // Write) has no effect unless the modified headers are
        // trailers.
```

Reply:
Yes, definitely, good catch. As the document implies, `WriteHeader()` and `Write()` are final (and calls to `WriteHeader()` seem to be intended to be called immediately before `Write()`). I will make the change.